## Problem 3: Prime Factors

The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600,851,475,143?

In [1]:
import time
import numpy as np
import math
import matplotlib
from matplotlib import pyplot as plt

matplotlib.rcParams['font.sans-serif'] = "Times New Roman"

In [2]:
''' 
Finds the prime factors of a number. The factors are found through recursion so the prime_factor_list should be 
passed to each level of recursion.

eg:
>>> prime_factors(1000, [])
>>> [5, 5, 5, 2, 2, 2]
'''

def prime_factors(number : int, prime_factor_list : list = []) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Finds the prime factors of the number and stores them in prime_factor_list. Returns this list to the user. It will
        search numbers until a factor is found, then divides the number by our factor and calls the function on the divided
        number. The list of factors is returned when the number is 1.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : number : The number for which the prime factors are to be found
    
        : list : prime_factor_list : The list of prime factors found. The factors are found through recursion so the
                                     prime_factor_list should be passed to each level of recursion. By default this list is
                                     empty and does not need to be passed to the function by the user.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : prime_factor_list : The list of prime factors found.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> prime_factors(1000)
        >>> [5, 5, 5, 2, 2, 2]
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(number, int): raise ValueError('Please enter an integer > 0 for the number argument.')
    if number < 1: raise ValueError('Please enter an integer > 0 for the number argument.')
    if not isinstance(prime_factor_list, list): raise ValueError('Please enter a list of integers for the prime_factor_list argument.')
    
    # Return list of factors when number is 1
    if number == 1:
        return prime_factor_list
    
    # Otherwise, seach for the next prime factor.
    # The smallest divisor will be prime since the prime will be smaller than any of its multiples.
    for factor in range(2, number+1):
        
        # Test if we have a factor.
        if number % factor == 0:
            
            # Call the function for the number divided by the factor.
            prime_factor_list = prime_factors(int(number/factor), prime_factor_list)
            
            # Add the factor to the list.
            prime_factor_list.append(factor)
            
            # Return the list
            return prime_factor_list

In [3]:
print(prime_factors(1000))

[5, 5, 5, 2, 2, 2]


In [4]:
test_number = 13195
start_time = time.time()
test_solution = prime_factors(test_number, [])
test_solution = np.max(test_solution)
end_time = time.time()

In [5]:
print(f"The largest prime factor of {test_number:,} is {test_solution:,}, computed in {end_time - start_time:.6f}s.")

The largest prime factor of 13,195 is 29, computed in 0.000000s.


In [6]:
problem_number = 600851475143 
start_time = time.time()
problem_solution_list = prime_factors(problem_number, [])
problem_solution = np.max(problem_solution_list)
end_time = time.time()

In [7]:
print(f"The prime factors of {problem_number:,} are {problem_solution_list}, the largest of which is {problem_solution:,}, computed in {end_time - start_time:.6f}s.")

The prime factors of 600,851,475,143 are [6857, 1471, 839, 71], the largest of which is 6,857, computed in 0.000496s.


In [8]:
# Can slightly modify function to only return the largest factor by returning the last factor found.

'''
Finds the largest prime factor of the number. It will search numbers until a factor is found, then divides the number 
by our factor and calls the function on the divided number. It stores the last factor as the largest prime factor.
Since we are searching up from 2, the last factor found will be the largest by construction.

eg. 
>>> find_largest_prime_factor(1000)
>>> 5
'''

def find_largest_prime_factor(number : int, largest_prime_factor : int = 1) -> int:
    '''
--- Function Description --------------------------------------------------------------------------------------------------
    Finds the largest prime factor of the number. It will search numbers until a factor is found, then divides the number 
    by our factor and calls the function on the divided number. It stores the last factor as the largest prime factor.
    Since we are searching up from 2, the last factor found will be the largest by construction.
---------------------------------------------------------------------------------------------------------------------------

--- Function Inputs -------------------------------------------------------------------------------------------------------
    : int : number : The number for which the prime factors are to be found

    : int : largest_prime_factor : The last of prime factor found. The factors are found through recursion so the
                                   largest_prime_factor should be passed to each level of recursion. This parameter 
                                   does not need to be passed to the function by the user.
---------------------------------------------------------------------------------------------------------------------------

--- Function Outputs ------------------------------------------------------------------------------------------------------
    : int : largest_prime_factor : The largest of prime factors found.
---------------------------------------------------------------------------------------------------------------------------

--- Function Examples -----------------------------------------------------------------------------------------------------
    >>> find_largest_prime_factor(1000)
    >>> 5
---------------------------------------------------------------------------------------------------------------------------
'''
    # Check types of function inputs:
    if not isinstance(number, int): raise ValueError('Please enter an integer > 0 for the number argument.')
    if not isinstance(largest_prime_factor, int): raise ValueError('largest_prime_factor argument has found non integer value.')
    
    # If the number is 1, then the last factor found was the last and therfore largest.
    if number == 1:
        return largest_prime_factor
    
    # Search through the numbers up from 2, by construction the first factor found will be the smallest.
    for factor in range(2, number+1):
        
        # Check if we have a factor of the number.
        if number % factor == 0:
            
            # Save the last factor found.
            largest_prime_factor = factor
            
            # Call the function with the number divided by the smallest factor.
            largest_prime_factor = find_largest_prime_factor(int(number/factor), largest_prime_factor)
            
            # Return the last factor found.
            return largest_prime_factor

In [9]:
problem_number = 600851475143 
start_time = time.time()
problem_solution = find_largest_prime_factor(problem_number)
end_time = time.time()

In [10]:
print(f"The largest prime factor of {problem_number:,} is {problem_solution:,}, computed in {end_time - start_time:.6f}s.")

The largest prime factor of 600,851,475,143 is 6,857, computed in 0.000496s.


### Problem 3 Solution: 

The largest prime factor of 600,851,475,143 is 6,857.